### PR TITLE
[FIX] point_of_sale: use last order time instead of system time when closing session

### PR DESCRIPTION
**Pull request by nim-odoo**
When the session is closed, the date used is the date of the last order instead of the
system date. This is necessary when the server is not on the same timezone than the user,

opw: 631497

### CODE REVIEW

File: addons/point_of_sale/point_of_sale.py [hunk 1]
```diff
@@ -515,7 +515,9 @@ def _confirm_orders(self, cr, uid, ids, context=None):
         account_move_obj = self.pool.get('account.move')
         pos_order_obj = self.pool.get('pos.order')
         for session in self.browse(cr, uid, ids, context=context):
-            local_context = dict(context or {}, force_company=session.config_id.journal_id.company_id.id)
+            last_order_id = sorted(session.order_ids.ids, reverse=True)[0]
+            last_order_date = pos_order_obj.browse(cr, uid, [last_order_id], context=context).date_order[:10]
+            local_context = dict(context or {}, force_company=session.config_id.journal_id.company_id.id, last_order_date=last_order_date)
             order_ids = [order.id for order in session.order_ids if order.state == 'paid']
 
             move_id = account_move_obj.create(cr, uid, {'ref' : session.name, 'journal_id' : session.config_id.journal_id.id, }, context=local_context)
```
Comment (beledouxdenis) [inline addons/point_of_sale/point_of_sale.py]: company_id = session.config_id.journal_id.company_id.id
local_context = dict(context or {}, force_company=company_id, company_id=company_id)
period_id = self.pool['account.period'].find(cr, uid, dt=session.start_at, context=local_context)
...
move_id = account_move_obj.create(cr, uid, {'ref' : session.name, 'journal_id' : session.config_id.journal_id.id, 'period_id': period_id}, context=local_context)

File: addons/point_of_sale/point_of_sale.py [hunk 2]
```diff
@@ -1037,6 +1039,7 @@ def _create_account_move_line(self, cr, uid, ids, session=None, move_id=None, co
 
         grouped_data = {}
         have_to_group_by = session and session.config_id.group_by or False
+        last_order_date = context['last_order_date'] if 'last_order_date' in context else None
```
Comment (beledouxdenis) [inline addons/point_of_sale/point_of_sale.py]: No need

File: addons/point_of_sale/point_of_sale.py [hunk 3]
```diff
@@ -1075,7 +1078,7 @@ def insert_data(data_type, values):
                 # if have_to_group_by:
 
                 sale_journal_id = order.sale_journal.id
-                period = account_period_obj.find(cr, uid, context=dict(context or {}, company_id=current_company.id))[0]
+                period = account_period_obj.find(cr, uid, dt=last_order_date, context=dict(context or {}, company_id=current_company.id))[0]
```
Comment (beledouxdenis) [inline addons/point_of_sale/point_of_sale.py]: No need as well. In fact, you can totaly remove the line 
period = account_period_obj.find(cr, uid, context=dict(context or {}, company_id=current_company.id))[0]

File: addons/point_of_sale/point_of_sale.py [hunk 4]
```diff
@@ -1463,6 +1466,16 @@ def create_from_ui(self, cr, uid, partner, context=None):
...
+    _inherit = 'account.move'
 
+    def create(self, cr, uid, vals, context=None):
+        if 'last_order_date' in context:
+            account_period_obj = self.pool.get('account.period')
+            period_id = account_period_obj.find(cr, uid, dt=context['last_order_date'], context=context)
+            result = super(account_move, self).create(cr, uid, dict(vals, date=context['last_order_date'], period_id=period_id[0]), context)
+        else:
+            result = super(account_move, self).create(cr, uid, vals, context)
+        return result
```
Comment (beledouxdenis) [inline addons/point_of_sale/point_of_sale.py]: No need at all either. Always try to never override create/write, that's not the best practice :).

**Comment by nim-odoo** [comment]
ok, updated

**Comment by beledouxdenis** [comment]
There is still one thing I am not sure of:
`period_id = self.pool['account.period'].find(cr, uid, dt=session.start_at[:10], context=local_context)`

I believe the dt param should be passed within user timezone, and session.start_at will be UTC tz, I think. You should therefore probably set it in user timezone before passing it. You should probably use context_timestamp to do that:
https://github.com/odoo/odoo/blob/8.0/openerp/osv/fields.py#L532

Something like
```
from datetime import datetime
...
start_at_datetime = datetime.strptime(session.start_at, tools.DEFAULT_SERVER_DATETIME_FORMAT)
date = fields.datetime.context_timestamp(cr, uid, start_at_datetime, context=local_context)
date = date.strftime(tools.DEFAULT_SERVER_DATE_FORMAT)
```

To check if you have to do that or not, simply set a pdb/pudb and check that session.start_at is indeed in UTC timezone and not within your timezone.

**Comment by nim-odoo** [comment]
It makes that dt should be passed in user tz since it is initialized as fields.date.context_today if empty. session.start_at is in UTC, so I updated the code.

**Comment by beledouxdenis** [comment]
Remove the comment
`# Convert time to user timezone`
We can easily understand what you do by reading the code, this comment is useless / redundant.

But then, it's fine for me :+1:

**Comment by nim-odoo** [comment]
ok

**Comment by nim-odoo** [comment]
Minor change following correction: move_id = self._create_account_move(cr, uid, **order.session_id.start_at**, order.name, order.sale_journal.id, order.company_id.id, context=context)

**Pull request merged**